manual/src/cmds/runtime-tracing.etex

```diff
@@ -70,6 +72,33 @@ The full set of events emitted by probes and their documentation can be found in
  section~\ref{Runtime_events}.
 \fi
 
+\subsection{s:runtime-tracing-custom-events}{Custom events}
+The runtime events system supports defining custom events. Identified by a
+globally unique name, they are emitted with payloads of built-in (\texttt{unit},
+\texttt{int}, \texttt{span}) and user-defined types. They are registered using
+the \texttt{Runtime_events.User.register} function. Once an event is registered,
+values for that event are emitted using \texttt{Runtime_events.User.write}.
+
+As consumer and producer can be different programs, they might not have the same
+set of registered events. The event name is used by the consumer to identify
+producer events, with a fallback mechanism for unknown events of built-in types.
+Unknown events for user-defined types are currently ignored, even if the type is
+defined both in the producer and consumer side.
+
+An extensible variant \texttt{tag} provides a way to attach additional data to
+events. The tag is not serialized, so it is only available for \textit{known}
+events.
+
+In summary, there are three cases for the consumer when an event is received:
+\begin{itemize}
+\item event is registered: payload and tag are available.
+\item event is not registered and has a built-in event type (unit, int, span):
+    only the payload is available.
+\item event is not registered and has a custom event type: event is dropped.
```

Comment:
Why is there a difference in behavior depending on the type of the payload? Wouldn't it make sense to drop all unregistered events?

Reply:
> Why is there a difference in behavior depending on the type of the payload? 

Due to how the event transport protocol is designed, the consumer can decode event payloads for built-in types only by using the event type ID field in the header. It is not currently possible for custom types, as the consumer has first to identify the event in order to obtain the decoder function. 

> Wouldn't it make sense to drop all unregistered events?

There is value in supporting unregistered events of built-in types. [runtime_events_tools](https://github.com/tarides/runtime_events_tools#trace)'s `olly trace` command was extended to integrate both GC events and unregistered events of the `span` type in a single trace (https://github.com/tarides/runtime_events_tools/pull/12/commits/5865899cafee1d1e50f90d0a4f55c6188dbf6cc3)

To me the other way around makes more sense: lift the limitation of unregistered events of custom types being dropped.

Reply:
> There is value in supporting unregistered events of built-in types. [We added support in olly trace.]

Can you spell this out for me, what is it useful for in practice?

Bonus question: Is it possible for an event to have several pieces of information as the payload, some custom and some built-in?

Reply:
>  what is it useful for in practice?

For example, I was trying to understand performance issues in a multicore eio program. Using olly and runtime events I figured out that on major GC, the domains had a hard time synchronizing (one domain was waiting up to 200ms) but I couldn't figure out why. By adding custom events of the span types in eio I was able to see that time was spent in a system call that was not marked as a blocking section. 

![screenshot_2023-02-15_at_15-20-23_eeww trace__1_mb__-_perfetto_ui](https://github.com/ocaml/ocaml/assets/966015/3c92725a-3107-44fd-8d90-06fc74b36395)

`olly trace` doesn't need to know about eio for that purpose.  In fact, any library using the `span` event could appear in the trace. To me, obtaining events from many sources in a single trace is very useful. 

> Bonus question: Is it possible for an event to have several pieces of information as the payload, some custom and some built-in?

Currently, that's not possible. One would have to create multiple events for that purpose.

Reply:
(and if I remember correctly, this is the offending C binding: https://github.com/ocaml-multicore/ocaml-uring/pull/86. Very hard to spot this bug without this tracing.)

Reply:
Thanks for the nice example! This is along the lines of: some listeners are only looking for specific events they know about, but some other listeners are trying to improve observability by reporting on all events, possibly in a degraded way.

Two thoughts on this:

1. It could be nice to suggest this situation in the manual (something like "Some generic listeners are used to visualize the trace of OCaml programs, and they will show your events [of standard payload] without modification".)
2. This example suggests that maybe dropping events with custom payload, instead of a best-effort approach of showing the event but nothing about the payload, is not the best approach. At least, in the current state of things, readers of the manual could be encouraged to systematically pair their custom-payload events with approximations of those with standard types that can be displayed, so that they can observe something with generic trace tools.

Reply:
To me, the paragraph added below this one is a good implementation of your suggestion 1.